pkg/lightning/backend/tidb/tidb.go

```diff
@@ -411,27 +479,81 @@ func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, colu
 		insertStmt.WriteByte(')')
 	}
 	insertStmt.WriteString(" VALUES")
+	return &insertStmt
+}
 
-	// Note: we are not going to do interpolation (prepared statements) to avoid
-	// complication arise from data length overflow of BIT and BINARY columns
+// Only used for TestWriteRowsErrorSkip
+var mockErrorCount = 0
 
-	for i, row := range rows {
-		if i != 0 {
-			insertStmt.WriteByte(',')
-		}
-		insertStmt.WriteString(string(row))
-	}
+func (be *tidbBackend) execStmts(ctx context.Context, stmtTasks []stmtTask, batch bool) error {
+stmtTaskLoop:
+	for _, stmtTask := range stmtTasks {
+	retryLoop:
+		for i := 0; i < writeRowsMaxRetryTimes; i++ {
+			stmt := stmtTask.stmt
+			txnCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			tx, err := be.db.BeginTx(txnCtx, &sql.TxOptions{})
```

Comment:
I still not understand why we need to depend on a transaction to exec these statements? If batch is true, then there should be only one statement, so tx is useless. If there are multi statements and some of them are expect to fail, then seems exec the statement one by one until meeting the first error seems to be a better choice, in this way wo can also avoid retry all the alredy succeeded statements again.